cmd/containerd-shim-runhcs-v1/task_hcs.go

```diff
@@ -859,3 +847,117 @@ func (ht *hcsTask) Stats(ctx context.Context) (*stats.Statistics, error) {
+	if ht.isWCOW {
+		return ht.updateWCOWResources(ctx, data, annotations)
+	}
+
+	return ht.updateLCOWResources(ctx, data, annotations)
+}
+
+func (ht *hcsTask) updateWCOWContainerCPU(ctx context.Context, cpu *specs.WindowsCPUResources) error {
+	// if host is 20h2+ then we can make a request directly to hcs
+	if osversion.Get().Build >= osversion.V20H2 {
+		req := &hcsschema.Processor{}
+		if cpu.Count != nil {
+			procCount := *cpu.Count
+			hostProcs := uint64(processorinfo.ProcessorCount())
+			if ht.host != nil {
+				hostProcs = uint64(ht.host.ProcessorCount())
+			}
+			if procCount > hostProcs {
+				log.G(ctx).WithFields(logrus.Fields{
+					"requested": procCount,
+					"assigned":  hostProcs,
+				}).Warn("Changing user requested CPUCount to current number of processors")
+				procCount = hostProcs
+			}
+			req.Count = int32(procCount)
+		}
+		if cpu.Maximum != nil {
+			req.Maximum = int32(*cpu.Maximum)
+		}
+		if cpu.Shares != nil {
+			req.Weight = int32(*cpu.Shares)
+		}
+		return ht.requestUpdateContainer(ctx, resourcepaths.SiloProcessorResourcePath, req)
+	}
+
+	return errdefs.ErrNotImplemented
+}
+
+func isValidWindowsCPUResources(c *specs.WindowsCPUResources) bool {
+	return (c.Count != nil && (c.Shares == nil && c.Maximum == nil)) ||
+		(c.Shares != nil && (c.Count == nil && c.Maximum == nil)) ||
+		(c.Maximum != nil && (c.Count == nil && c.Shares == nil))
+}
+
+func (ht *hcsTask) updateWCOWResources(ctx context.Context, data interface{}, annotations map[string]string) error {
+	resources, ok := data.(*specs.WindowsResources)
+	if !ok {
+		return errors.New("must have resources be type *WindowsResources when updating a wcow container")
+	}
+	if resources.Memory != nil && resources.Memory.Limit != nil {
+		newMemorySizeInMB := *resources.Memory.Limit / bytesPerMB
+		memoryLimit := normalizeContainerMemorySize(ctx, ht.id, newMemorySizeInMB)
+		if err := ht.requestUpdateContainer(ctx, resourcepaths.SiloMemoryResourcePath, memoryLimit); err != nil {
+			return err
+		}
+	}
+	if resources.CPU != nil {
+		if !isValidWindowsCPUResources(resources.CPU) {
+			return fmt.Errorf("invalid cpu resources request for container %s: %v", ht.id, resources.CPU)
+		}
+		if err := ht.updateWCOWContainerCPU(ctx, resources.CPU); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (ht *hcsTask) updateLCOWResources(ctx context.Context, data interface{}, annotations map[string]string) error {
+	resources, ok := data.(*specs.LinuxResources)
+	if !ok || resources == nil {
+		return errors.New("must have resources be non-nil and type *LinuxResources when updating a lcow container")
+	}
+	settings := guestrequest.LCOWContainerConstraints{
+		Linux: *resources,
+	}
+	return ht.requestUpdateContainer(ctx, "", settings)
+}
+
+func (ht *hcsTask) requestUpdateContainer(ctx context.Context, resourcePath string, settings interface{}) error {
+	var modification interface{}
+	if ht.isWCOW {
+		modification = &hcsschema.ModifySettingRequest{
```

Comment:
Should we be using a `guestrequest.GuestRequest` for a hypervisor-isolated Windows container? Or does this request work for all Windows containers (process and hypervisor isolated)?

Reply:
`hcsschema.ModifySettingRequest` can be used to issue a modify request to a hypervisor isolated container. In particular, we need to provide the container resource path `ResourcePath`.

Reply:
Okay, so this works for both hypervisor and process isolated Windows container?

Reply:
yes